### PR TITLE
fix(gatsby-plugin-layout): Add missing HMR for layout component

### PR DESCRIPTION
**Pull request by cometkim**
## Description

It seems HMR doesn't work when the layout component changed

I added `hot` wrapper follow this code:

https://github.com/gatsbyjs/gatsby/blob/76ac266a19502c5c038a5fab697dcbfaed5ec664/packages/gatsby/src/bootstrap/requires-writer.js#L128-L142

### CODE REVIEW

**Comment by pieh** [comment]
This isn't that straight forward unfortunately - this will add `hot` for production builds as well :( we need to conditionally add `hot` wrapper and `react-hot-loader` import only if we are in development mode

**Comment by cometkim** [comment]
Oh, @pieh you right. I forgot that even I wrote some [macro](https://github.com/cometkim/react-hot-reload.macro/blob/master/index.js) for it. :sweat_smile: 

Fixed it allows for `process.env.NODE_ENV !== "production"` expect it to be eliminated by webpack in production builds.

File: packages/gatsby-plugin-layout/src/wrap-page.js [hunk 1]
```diff
@@ -4,6 +4,10 @@ const preferDefault = m => (m && m.default) || m
 let Layout
 try {
   Layout = preferDefault(require(GATSBY_LAYOUT_COMPONENT_PATH))
+  if (process.env.NODE_ENV !== `production`) {
```
Comment (pranshuchittora) [inline packages/gatsby-plugin-layout/src/wrap-page.js]: !prod looks confusing, how about `=== dev`

Comment (cometkim) [reply packages/gatsby-plugin-layout/src/wrap-page.js]: Fixed, thanks for your feedback @pranshuchittora

File: packages/gatsby-plugin-layout/src/wrap-page.js [hunk 2]
```diff
@@ -4,6 +4,10 @@ const preferDefault = m => (m && m.default) || m
 let Layout
 try {
   Layout = preferDefault(require(GATSBY_LAYOUT_COMPONENT_PATH))
+  if (process.env.NODE_ENV === `develeopment`) {
```
Comment (sidharthachatterjee) [inline packages/gatsby-plugin-layout/src/wrap-page.js]:
```suggestion
  if (process.env.NODE_ENV === `development`) {
```

**Comment by cometkim** [comment]
@sidharthachatterjee 

I've tested it by manipulating my node_modules directly ant it was works. I'll investigate more it in this weekend.

**Comment by cometkim** [comment]
Peek 2019-11-14 06-34

@sidharthachatterjee Fixed it!

**Comment by wardpeet** [comment]
Hey thanks for creating the PR. I have a few global comments.

Can you run `yarn add react-hot-loader` as a dependency as we might get hoisting problems in the future if we don't.

The documentation points out that hot should be used on export level instead of import level.

<details>
<summary>See https://github.com/gaearon/react-hot-loader#important</summary>

**!!** Use `hot` only for module `exports`, not for module `imports`. **!!**

```js
import { hot } from 'react-hot-loader/root';

const App = () => 'Hello World!';

export default hot(App);
```

Keep in mind - by importing `react-hot-loader/root` you are setting up a boundary for update event propagation.

The higher(in module hierarchy) you have it - the more stuff would be updated on Hot Module Replacement.

To make RHL more reliable and safe, please place `hot` _below_ (ie somewhere in _imported_ modules):

* react-dom
* redux store creation
* any data, you want to preserve between updates
* big libraries

You may(but it's not required) place `hot` to the every route/page/feature/lazy chunk, thus make updates more scoped.

You don't need to wrap every component with `hot`, application work work fine with a single one.
</details>

In https://github.com/gatsbyjs/gatsby/pull/18949/files#diff-c4777e791f64fbb6c82255341b381dab, we use a loader called `react-hot-loader-loader` to accomplish this. You can add a webpack rule for the layout file that is used inside the component property.

**Comment by LekoArts** [comment]
@cometkim Hi! 👋 Do you have time to address the requested changes?

**Comment by cometkim** [comment]
Hey @LekoArts, I just changed to using react-hot-loader-loader as @wardpeet's suggestion, and it seems to work well.

**Comment by cometkim** [comment]
Could anyone run `yarn workspace add gatsby-plugin-layout react-hot-loader-loader react-hot-loader` and push lock file? I'm facing some problem to add dependencies.

**Comment by pieh** [comment]
> Could anyone run yarn workspace add gatsby-plugin-layout react-hot-loader-loader react-hot-loader and push lock file? I'm facing some problem to add dependencies.

I'm taking second look now, so will do this as well

**Comment by pieh** [comment]
I still do get reloads. I added some logs to webpack rule test and I never return true there:

```
success extract queries from components - 0.096s
testing false /Users/misiek/test/pr19056/node_modules/@babel/runtime/helpers/esm/inheritsLoose.js /Users/misiek/test/pr19056/src/layouts/index
testing false /Users/misiek/test/pr19056/node_modules/react/index.js /Users/misiek/test/pr19056/src/layouts/index
testing false /Users/misiek/test/pr19056/node_modules/webpack/buildin/harmony-module.js /Users/misiek/test/pr19056/src/layouts/index
success Re-building development bundle - 0.688s
```

so:
 1) `absoluteComponentPath` (at least default one), won't ever match actual file it seems as it lacks the extension
 2) even if it matched - we still don't test for layout file with this setup

I'm not quite sure how you got it to work honestly

As a note:
manually adding:
```
import { hot } from "react-hot-loader"

export default hot(module)(LayoutComponent)
```

to layout component make it work (this is what the `react-hot-loader-loader` does essentially), but this is not what we want users to do

**Comment by pvdz** [comment]
@cometkim Hi! This PR is pretty old and hasn't seen an update in a while. Is it still relevant? Can you let us know what your plan is with this PR? Perhaps it should be in Draft mode? Thanks :)

**Comment by pieh** [comment]
I was thinking more about this PR lately:
with `react-refresh` being supported in core (under conditional flag) doing any kind of `react-hot-loader` handling in plugins in unsafe. We should likely wrap `wrapRootElement`/`wrapPageElement` in core runtime in something like this, but conditionally depending if we use `react-refresh` or `react-hot-loader` - similar to what we already do in `sync-requires` writer: https://github.com/gatsbyjs/gatsby/blob/8ee74c1f85f75eb7ede2c22eb87eee21ebc9e8c6/packages/gatsby/src/bootstrap/requires-writer.js#L148-L170

So I will close the PR because this is not right direction (it wasn't right from the start, as other plugins would need similar approach as well, just `react-refresh` made me realize that)

**Pull request closed without merging**